src/ol_infrastructure/components/aws/cache.py

```diff
@@ -0,0 +1,211 @@
+"""This module defines a Pulumi component resource for enforcing best practices for building AWS Elasticache clusters.
+
+This includes:
+- Create a parameter group for each deployed cluster
+- Create a clustered deployment of Redis or Memcached
+"""
+import re
+
+from typing import Any, Dict, List, Optional, Text, Union, cast
+
+import pulumi
+
+from pulumi_aws import elasticache
+from pydantic import PositiveInt, conint, validator
+
+from ol_infrastructure.lib.aws.elasticache_helper import (
+    cache_engines,
+    parameter_group_family
+)
+from ol_infrastructure.lib.ol_types import AWSBase
+
+
+class OLAmazonCacheConfig(AWSBase):
+    encrypt_transit: Optional[bool] = None
+    encrypted: Optional[bool] = None
+    auth_token: Optional[Text] = None
+    shard_count: Optional[int] = None
+    kms_key_id: Optional[Text] = None
+    snapshot_retention_days: Optional[int] = None
+    auto_upgrade: bool = True  # Automatically perform ugprades of minor versions
+    apply_immediately: bool = False
+    cluster_description: Text
+    cluster_name: Text
+    engine: Text
+    engine_version: Text
+    instance_type: Text
+    num_instances: PositiveInt = PositiveInt(3)
+    parameter_overrides: Optional[Dict[Text, Any]] = None
+    port: int
+    security_groups: List[Text]
+    subnet_group: Union[Text, pulumi.Output[str]]  # the name of the subnet group created in the OLVPC component
+
+    class Config:  # noqa: WPS431, WPS306, D106
+        arbitrary_types_allowed = True
+
+    @validator('engine')
+    def is_valid_engine(cls: 'OLAmazonCacheConfig', engine: Text) -> Text:  # noqa: N805, D102
+        valid_engines = cache_engines()
+        if engine not in valid_engines:
+            raise ValueError('The specified cache engine is not a valid option in AWS.')
+        return engine
+
+    @validator('engine_version')
+    def is_valid_version(  # noqa: D102
+            cls: 'OLAmazonCacheConfig',  # noqa: N805
+            engine_version: Text,
+            values: Dict  # noqa: WPS110
+    ) -> Text:
+        engine = values.get('engine')
+        engines_map = cache_engines()
+        if engine_version not in engines_map.get(engine, []):
+            raise ValueError(f'The specified version of the {engine} engine is not supported in AWS.')
+        return engine_version
+
+
+class OLAmazonRedisConfig(OLAmazonCacheConfig):
+    cluster_mode_enabled: bool = True
+    encrypt_transit: bool = True
+    auth_token: Optional[Text] = None
```

Comment:
Wouldn't this throw a ValueError since its value is `None` and `encrypt_transit` is `True`?

Reply:
If you use the defaults, then yes. You can solve that situation by either setting `encrypt_transit` to False or by setting the `auth_token` value.